### PR TITLE
Updated NetworkTopology configuration file format

### PR DESCRIPTION
**Pull request by bolt12**
new file format follows the following structure:

```haskell
data RootAddress = RootAddress
  { addrs :: [RelayAddress]
  , advertise :: PeerAdvertise
  } deriving (Eq, Show)

data LocalRootPeers = LocalRootPeers
  { localRoots :: RootAddress
  , valency :: Int
  } deriving (Eq, Show)

newtype LocalRootPeersGroups = LocalRootPeersGroups
  { groups :: [LocalRootPeers]
  } deriving (Eq, Show)

newtype PublicRootPeers = PublicRootPeers
  { publicRoots :: RootAddress
  } deriving (Eq, Show)

data NetworkTopology = RealNodeTopology !LocalRootPeersGroups ![PublicRootPeers] !UseLedger
  deriving (Eq, Show)
```

### CODE REVIEW

**Comment by bolt12** [comment]
Needs this ouroboros-network PR to compile: https://github.com/input-output-hk/ouroboros-network/pull/3069

**Comment by bolt12** [comment]
@Jimbo4350  Thank you! That commit seems to be an artifact of @karknu force-push, I still need to rebase accordingly

**Review by Jimbo4350 — COMMENTED** [review]
LGTM! However the topology files & docs need to be updated. QA  & devops also need to be informed that the topology file format will have breaking changes. cc: @disassembler @dorin100 @mkoura @rvl

**Comment by bolt12** [comment]
Thank you @Jimbo4350, still need to make sure the other PR lands on `ouroboros-network` and then I'll also update the tag

File: doc/getting-started/understanding-config-files.md [hunk 1]
```diff
@@ -6,21 +6,55 @@ Tells your node to which nodes in the network it should talk to. A minimal versi
...
 ```json
 {
-  "Producers": [
+  "LocalProducers": {
+    "groups": [
+      {
+        "localRoots": {
+          "addrs": [
+            {
+              "isDomain": true,
```
Comment (karknu) [inline doc/getting-started/understanding-config-files.md:15]: Can you get rid of 'isDomain'?  the previous config file didn't require it I don't see why the new format makes it necessary.

File: doc/getting-started/understanding-config-files.md [hunk 2]
```diff
@@ -6,21 +6,55 @@ Tells your node to which nodes in the network it should talk to. A minimal versi
...
     }
   ]
 }
 ```
-* This means that your node will contact node at ip `x.x.x.x` on `port 3001`.
 
-* `valency` tells the node how many connections your node should have. It only has an effect for dns addresses. If a dns address is given, valency governs to how many resolved ip addresses should we maintain active (hot) connection; for ip addresses, valency is used as a Boolean value, where `0` means to ignore the address.
+* The main difference between `LocalProducers` and `PublicProducers` is that with the former
+    you can specify different groups with different valencies. That can be useful to
+    inform your node of different targets withing a group to achieve.
```
Comment (karknu) [inline doc/getting-started/understanding-config-files.md:47]: This only explains the difference between `LocalProducers` and `PublicProducers` from a json format point of view. `localProduers` is for peers which the node always should have as hot, such as their own block producer. `PublicProducers` represent a source of fallback peers, a source of peers to be used if peers from the ledger is disabled or unavailable.

Comment (bolt12) [reply doc/getting-started/understanding-config-files.md:47]: Thank you, I'll improve this description!

**Comment by bolt12** [comment]
Example config:

```
{
  "LocalProducers": {
    "groups": [
      {
        "localRoots": {
          "addrs": [
            {
              "addr": "relays-new.cardano-mainnet.iohk.io",
              "port": 3001
            }
          ],
          "advertise": false
        },
        "valency": 1
      }
    ]
  },
  "PublicProducers": [
    {
      "publicRoots" : {
        "addrs": [
          {
            "addr": "35.180.198.31",
            "port": 7776
          }
        ],
        "advertise": false
      }
    }
  ],
  "useLedgerAfterSlot": false
}
```

File: doc/getting-started/understanding-config-files.md [hunk 3]
```diff
@@ -6,21 +6,55 @@ Tells your node to which nodes in the network it should talk to. A minimal versi
...
+            "addr": "y.y.y.y",
+            "port": 3002
+          }
+        ],
+        "advertise": false
+      }
     }
-  ]
+  ],
+  "useLedgerAfterSlot": false
```
Comment (karknu) [inline doc/getting-started/understanding-config-files.md:38]: useLedgerAfterSlot isn't a bool , it is a slot number. Use a number less than 0 to disable.

Comment (bolt12) [reply doc/getting-started/understanding-config-files.md:38]: oops my bad!

**Review by coot — CHANGES_REQUESTED** [review]

File: doc/getting-started/understanding-config-files.md [hunk 4]
```diff
@@ -6,21 +6,55 @@ Tells your node to which nodes in the network it should talk to. A minimal versi
...
-Your __block-producing__ node must __ONLY__ talk to your __relay nodes__, and the relay node should talk to other relay nodes in the network. Go to our telegram channel to find out IP addresses and ports of peers.
+* This means that your node will contact node at ip `x.x.x.x` on `port 3001`, and resolve
+    dns domain `y.y.y.y` (assuming they are), and try to maintain a connection with at least `1` of the
+    resolved ips.
 
+* `isDomain` tells if a given address is a dns address or not. If true and object `domain: { domain, port }` should be specified instead.
+
+* `valency` tells the node how many connections your node should try to pick from the given group. If a dns address is given, valency governs to how many resolved ip addresses should we maintain active (hot) connection.
+
+Your __block-producing__ node must __ONLY__ talk to your __relay nodes__, and the relay node should talk to other relay nodes in the network. Go to our telegram channel to find out IP addresses and ports of peers.
```
[the remark this answers is not in the document]
Comment (bolt12) [reply doc/getting-started/understanding-config-files.md:57]: I'll remove the last sentence then! Thank you!

**Comment by bolt12** [comment]
@coot for CI to fix this needs #2647

**Review by coot — APPROVED** [review]

**Review by karknu — APPROVED** [review]
LGTM

**Pull request merged**